Add new Protobuf-based MessageBus DocumentAPI protocol

vekterli opened the pull request:
@geirst please review schema/C++ code (primary focus on `routable_factories_8.cpp`), @jonmv please review schema/Java code (primary focus on `RoutableFactories80.java` and its codec design). Note that this commit by itself will **fail** its PR-build due to missing cross-language test files. This is intentional, as these are tied to the protocol version and will be added as a separate commit.

This adds an entirely new implementation of the internal MessageBus DocumentAPI protocol, which shall be functionally 1-to-1 compatible with the existing legacy protocol.

> [!CAUTION]
> **MUST NOT** be merged before a concrete Vespa release version (in the future) has been assigned to the protocol

New protobuf schemas have been added to the top-level documentapi module, which are separated into different domains of responsibility:

 * CRUD messages
 * Visiting messages
 * Data inspection messages

As well as a schema for shared, common message types.

Both C++ and Java protocol implementations separate serialization and deserialization into a codec abstraction per message type, which hides the boilerplate required for Protobuf buffer management. The Java version is a tad more verbose due to generics type-erasure.

This protocol does _not_ currently support lazy (de-)serialization in Java, as the existing mechanisms for doing so are inherently tied to the legacy protocol version. Performance tests will decide if we need to introduce such functionality to the new protocol version.

To avoid having the new protocol go live in production, this commit changes the semantics of how MessageBus version reporting works (at least for the near future); instead of reporting the current Vespa _release_ version, it reports the highest supported _protocol_ version. This lets us conditionally enable the new protocol by reporting a MessageBus version greater than or equal to the protocol version _iff_ the protocol should be active. This is currently done with a dedicated (presumably short-lived) environment variable.

The new protocol is disabled by default.

Other changes:
 * Protocol tests have been moved up one package directory level to be aligned with the actual package of the classes they test. This allows for using package-protected constructors in the serialization tests.
 * `DocumentDeserializer` now exposes the underlying document type repo/manager. This is done to detangle `Document`/`DocumentUpdate` deserialization from the underlying wire buffer management.
 * `RemoveLocationMessage` at long last contains a bucket space, which was forgotten when we initially added this concept to the other messages, and where the pain of adding it in later was too big (not so anymore!).

Unit tests for both C++ and Java have been hoisted from the legacy test suite, cleaned up and extended with additional cases. The C++ tests use the old unit test kit and should receive a good follow-up washing and GTest-rewrite.

**Important**: due to how MessageBus protocol versioning works, the final protocol version is _not_ yet decided, as setting it requires syncing against our build systems. A follow-up commit will assign the final version as well as include all required binary test files.

vekterli committed:
Add new Protobuf-based MessageBus DocumentAPI protocol

This adds an entirely new implementation of the internal MessageBus
DocumentAPI protocol, which shall be functionally 1-to-1 compatible
with the existing legacy protocol.

New protobuf schemas have been added to the top-level documentapi
module, which are separated into different domains of responsibility:

 * CRUD messages
 * Visiting messages
 * Data inspection messages

As well as a schema for shared, common message types.

Both C++ and Java protocol implementations separate serialization and
deserialization into a codec abstraction per message type, which
hides the boilerplate required for Protobuf buffer management. The
Java version is a tad more verbose due to generics type-erasure.

This protocol does _not_ currently support lazy (de-)serialization
in Java, as the existing mechanisms for doing so are inherently tied
to the legacy protocol version. Performance tests will decide if we
need to introduce such functionality to the new protocol version.

To avoid having the new protocol go live in production, this commit
changes the semantics of how MessageBus version reporting works (at
least for the near future); instead of reporting the current Vespa
_release_ version, it reports the highest supported _protocol_ version.
This lets us conditionally enable the new protocol by reporting a
MessageBus version greater than or equal to the protocol version _iff_
the protocol should be active.

The new protocol is disabled by default.

Other changes:
 * Protocol tests have been moved up one package directory level to
   be aligned with the actual package of the classes they test. This
   allows for using package-protected constructors in the serialization
   tests.
 * `DocumentDeserializer` now exposes the underlying document type
   repo/manager. This is done to detangle `Document`/`DocumentUpdate`
   deserialization from the underlying wire buffer management.
 * `RemoveLocationMessage` at long last contains a bucket space,
   which was forgotten when we initially added this concept to the
   other messages, and where the pain of adding it in later was
   too big (not so anymore!).

Unit tests for both C++ and Java have been hoisted from the legacy
test suite, cleaned up and extended with additional cases. The C++
tests use the old unit test kit and should receive a good follow-up
washing and GTest-rewrite.

**Important**: due to how MessageBus protocol versioning works, the
final protocol version is _not_ yet decided, as setting it requires
syncing against our build systems. A follow-up commit will assign
the final version as well as include all required binary test files.

jonmv commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java:
```diff
@@ -0,0 +1,931 @@
+            if (encoderFn != null) {
+                throw new IllegalArgumentException("Encoder already set");
+            }
+            encoderFn = fn;
```
Could consider `requireNonNull(fn)` as well, to be in line with the above check. Likewise for the `decoder(...)` below.

jonmv replied:
I see the nullity is guarded against in `build()`, so this is less important.

vekterli replied:
Since this particular API is explicitly internal use only I'll keep this as-is for now

jonmv commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java:
```diff
@@ -0,0 +1,931 @@
+            return this;
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> decoder(ProtoT messagePrototype, Function<ProtoT, DocApiT> fn) {
```
Did you consider just passing the `parser()` here, instead of the `getDefaultInstance()`? It's possible this is the idiomatic way with protobuf, but that's not obvious to me. That also lets you get rid of the cast on L108. 
Similarly for `decoderWithRepo(...)`.

vekterli replied:
Have changed the code to use the underlying parser directly

jonmv commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java:
```diff
@@ -0,0 +1,931 @@
+                try {
+                    var protoObj = messagePrototype.getParserForType().parseFrom(buf.getBuf().getByteBuffer());
+                    return fn.apply(protoClass.cast(protoObj));
+                } catch (IOException e) {
```
Are we guaranteed the `IOException` is never thrown?

jonmv replied:
Yes, but with no trace of what went wrong?

vekterli replied:
Good point. I see there is a higher-level `decode` method which does at least print out the exception message as a warning. Will runtime-wrap and re-throw instead of just returning null

jonmv replied:
I think that's a good idea.

vekterli replied:
Now has explicit catching and logging for both encoding and decoding

jonmv commented on documentapi/src/test/java/com/yahoo/documentapi/messagebus/protocol/Messages80TestCase.java:
```diff
@@ -0,0 +1,729 @@
+        @Override
+        public void run() {
+            var msg = new PutDocumentMessage(new DocumentPut(new Document(protocol.getDocumentTypeManager().getDocumentType("testdoc"), "id:ns:testdoc::")));
+            msg.setTimestamp(666);
```
👹

jonmv commented on documentapi/src/test/java/com/yahoo/documentapi/messagebus/protocol/Messages80TestCase.java:
```diff
@@ -0,0 +1,729 @@
+            update.addFieldPathUpdate(new RemoveFieldPathUpdate(docType, "intfield", "testdoc.intfield > 0"));
+
+            var msg = new UpdateDocumentMessage(update);
+            msg.setNewTimestamp(777);
```
👼

jonmv commented on messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCNetwork.java:
```diff
@@ -301,15 +300,34 @@ public boolean destroy() {
+        // Disallow-version MUST be lower than that used as a protocol lower bound in
+        // DocumentProtocol.java and the exact same as that used in C++ for the same purposes.
+        // ... Or else!
+        // TODO remove this glorious hack once protobuf protocol is enabled by default
```
🙈

vekterli replied:
👀 💦

jonmv replied:
Will this usage of `protobuf` be sufficient to never need the "version negotiation" again? I'm not super-familiar with this glorious new technology. I mean, where does _it_ check protocol versions, or whatever it does.

vekterli replied:
Also, enabling the Protobuf protocol by default would just change the reported version to always be one that is >= the version that triggers both sides to use the correct codecs. The same dance could then be done again later if we wanted to add a new protocol with a version > the Protobuf protocol version.

jonmv reviewed (commented):
This is great! I just have minor comments.

geirst commented on documentapi/src/vespa/documentapi/messagebus/routable_factories_8.h:
```diff
@@ -0,0 +1,72 @@
+
+namespace documentapi::messagebus {
+
+class RoutableFactories80 {
```
Consider adding class comment.

vekterli replied:
Done

geirst commented on documentapi/src/vespa/documentapi/messagebus/routable_factories_8.cpp:
```diff
@@ -0,0 +1,883 @@
+
+document::GlobalId get_global_id(const protobuf::GlobalId& src) {
+    if (src.raw_gid().size() != document::GlobalId::LENGTH) {
+        throw document::DeserializeException("Unexpected serialized protobuf GlobalId size");
```
Consider adding actual and expected lengths to exception message.

vekterli replied:
Done

geirst reviewed (dismissed):
C++ code changes 👍

vekterli committed:
Improve codec error reporting and use Protobuf type parsers directly

Errors during (de-)serialization will now be logged, alongside the
message type in question. Due to differences in internal wiring, the
Java version logs the Document API type whereas C++ logs the protobuf
message type (and also has throttled logging).

geirst reviewed (dismissed):
👍

vekterli committed:
Tag Protobuf protocol boundary version and add binary test files

All _reported_ versions >= 8.310 use Protobuf protocol, all lower
versions use the legacy protocol. Reported version is controlled
by an environment variable and defaults to 8.309, i.e. the legacy
version.

jonmv commented:
🍾
